packages/node-core/src/indexer/connectionPoolState.manager.ts

```diff
@@ -0,0 +1,269 @@
+// Copyright 2020-2022 OnFinality Limited authors & contributors
+// SPDX-License-Identifier: Apache-2.0
+
+import chalk from 'chalk';
+import {toNumber} from 'lodash';
+import {IApiConnectionSpecific, errorTypeToScoreAdjustment} from '..';
+import {getLogger} from '../logger';
+import {ApiErrorType} from './connectionPool.service';
+
+const RETRY_DELAY = 60 * 1000;
+const MAX_FAILURES = 5;
+const RESPONSE_TIME_WEIGHT = 0.7;
+const FAILURE_WEIGHT = 0.3;
+
+export interface ConnectionPoolItem<T> {
+  endpoint: string;
+  performanceScore: number;
+  backoffDelay: number;
+  failureCount: number;
+  rateLimited: boolean;
+  failed: boolean;
+  lastRequestTime: number;
+  connected: boolean;
+  timeoutId?: NodeJS.Timeout;
+}
+
+const logger = getLogger('connection-pool-state');
+
+export interface IConnectionPoolStateManager<T extends IApiConnectionSpecific<any, any, any>> {
```

Comment:
I think this interface should be much simpler. We only really need what would be called from the workers. To me this would be getting the state and updating the state.

E.g. Im not sure why these are in the interface:
* `shutdown`
*  `deleteFromPool`
*  `addToConnections`

Reply:
they get called from `ConnectionPoolService` in workers

Reply:
That doesn't seem right to me. Maybe more logic needs to be in connection pool state manager.

To me only the main thread should decide if connections are added or removed.

Reply:
we keep all the apis in workers because that is where the fetching happens. so, some operations like attempting to reconnect to endpoints will be done in workers, add/delete will be called in the host pool state manager depending on the result of reconnection in workers.